### PR TITLE
Feature/sg merge stats

### PR DESCRIPTION
**Pull request by byb121**
Changes:

* tidied up code a bit in returning error messages when checking files.
* added an option to output stats when merging single guided read counts.

### CODE REVIEW

File: crispr_read_counts/single_guide_merge.py [hunk 1]
```diff
@@ -65,3 +88,21 @@ def get_counts_results(in_f: TextIO, with_or_without_plasmid_counts: bool):
...
+    'zero_count_guides': 0,
+    'low_count_guides': 0,
+    'total_counts': 0
+  }
+
+  for a_count in sample_count.values():
+    stats['total_counts'] += a_count
+    if a_count == 0:
+      stats['zero_count_guides'] += 1
+    if a_count < SingleGuideReadCounts.LOW_COUNT_GUIDES_THRESHOLD:
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_merge.py:104]: An `elif` here would save some nanoseconds.

File: crispr_read_counts/single_guide_count.py [hunk 2]
```diff
@@ -14,30 +14,27 @@
...
-  if args['ref'] and not check_file_readable(args['ref']):
-    sys.exit(error_msg('Provided reference file does not exist or have no permission to read: %s' % args['ref']))
-  if args['plasmid'] and not check_file_readable(args['plasmid']):
-    sys.exit(error_msg('Provided plasmid count file does not exist or have no permission to read: %s' % args['plasmid']))
-  if not check_file_writable(args['output']):
-    sys.exit(error_msg('Cannot write to provided output count file: %s' % args['output']))
-  if args['stats'] and not check_file_writable(args['stats']):
-    sys.exit(error_msg('Cannot write to provided output stats file: %s' % args['stats']))
+def check_files(args: Dict[str, Any]):
+  for file_type, file_path in zip(['library', 'input'], [args['library'], args['input']]):
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: I'd propose:

```python
for file_type in ['library', 'input']:
  file_path: str = args[file_type]
  # check_file_readable(file_path, ...)
```

File: crispr_read_counts/utils.py [hunk 3]
```diff
@@ -43,25 +44,38 @@ def open_plain_or_gzipped_file(file: str):
...
 
-def check_file_readable(fn):
+def check_file_readable(fn, msg_if_fail=None):
+  result = True
   if os.path.isfile(fn):
-    return os.access(fn, os.R_OK)
+    result = os.access(fn, os.R_OK)
   else:
-    return False
+    result = False
```
Comment (lbarbon) [inline crispr_read_counts/utils.py]: To save a couple of statements, I'd propose:

```python
result: bool = os.path.isfile(fn) and os.access(fn, os.R_OK)
```

File: crispr_read_counts/single_guide_merge.py (hunk 1 above)
Comment (byb121) [reply crispr_read_counts/single_guide_merge.py:104]: low_count_guide should include zero_count_guide, `elif` will exclude them.

Comment (lbarbon) [reply crispr_read_counts/single_guide_merge.py:104]: I see, okay then.

File: crispr_read_counts/utils.py [hunk 4]
```diff
@@ -43,25 +44,38 @@ def open_plain_or_gzipped_file(file: str):
...
-      return os.access(fn, os.W_OK)
+      result = os.access(fn, os.W_OK)
+    else:
+      result = False  # path is a dir, so cannot write as a file
+  else:  # target does not exist, check perms on parent dir
+    pdir = os.path.dirname(fn)
+    if not pdir:
+      pdir = '.'
+    # target is creatable if parent dir is writable
+    result = os.access(pdir, os.W_OK)
```
Comment (lbarbon) [inline crispr_read_counts/utils.py]: This should be equivalent:

```python
result: bool = (
  (os.path.isfile(fn) and os.access(fn, os.W_OK)) if os.path.exists(fn) else
  os.access(os.path.dirname(fn) or '.', os.W_OK)
)
```

If the user mistypes the file name the check will fall back to evaluating the directory, though. Is that okay?

Comment (byb121) [reply crispr_read_counts/utils.py]: I've been thinking about what those lines can be :)

I thought those "if else" looked a bit clearer, so what's the selling point of your line?

Comment (lbarbon) [reply crispr_read_counts/utils.py]: The nested conditional blocks could perform any action, so you have to read through all of them to understand what they affect; an expression can only return a value, so the purpose is clearer at first glance.

Comment (byb121) [reply crispr_read_counts/utils.py]: You're right. The point is taken :)

**Pull request merged**